tests/gpu-tests/test_context_retry.py

```diff
@@ -0,0 +1,274 @@
+    # TODO: Currently this is just a single turn message. Need to add tests for multi-turn messages.
+    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+        try:
+            for _ in range(num_samples):
+                # Create large prompt that will likely exceed context limits
+                large_prompt = {"question": "a" * 500_000 + "b" * 500_000}
+                temp_file.write(json.dumps(large_prompt).encode())
+                temp_file.write(b"\n")
+            temp_file.flush()
+
+            # Copy to Docker container
+            docker_run(f"cp {temp_file.name} {input_file}")
+        finally:
+            # Clean up temporary file
+            os.unlink(temp_file.name)
+
+
+class ContextRetryTestSuite:
+    """Main test suite for context retry functionality"""
+
+    def __init__(self):
+        self.env = TestEnvironment()
+        self.env.validate_environment()  # validate environment before initializing other objects
+        self.config = TestConfig()
+        self.cmd_builder = CommandBuilder(self.env, self.config)
+        self.output_manager = OutputManager()
+        self.validator = MetricsValidator(self.config)
+
+    def run_eval_test(self, test_name: str, enable_soft_fail: bool, retry_strategy: str, expect_success: bool = True):
+        """Run evaluation test with specified parameters"""
+        output_dir = self.output_manager.setup_output_dir(self.env.model_type, test_name)
+        cmd = self.cmd_builder.build_eval_cmd(output_dir, enable_soft_fail, retry_strategy)
+
+        subprocess.run(cmd, shell=True, check=True)
+
+        if expect_success:
+            return self.validator.validate_eval_metrics(output_dir)
+        else:
+            return self.validator.validate_eval_failure(output_dir)
+
+    def run_eval_completion_test(self, test_name: str, enable_soft_fail: bool, retry_strategy: str):
+        """Run evaluation test for completion."""
+        output_dir = self.output_manager.setup_output_dir(self.env.model_type, test_name)
+        cmd = self.cmd_builder.build_eval_cmd(output_dir, enable_soft_fail, retry_strategy)
+        subprocess.run(cmd, shell=True, check=True)
+
+        self.validator.validate_eval_completion_but_empty_generation(output_dir)
+
+    def run_generation_test(self, test_name: str, retry_strategy: str):
+        """Run generation test with specified retry strategy"""
+        output_dir = self.output_manager.setup_output_dir(self.env.model_type, test_name)
+        input_file, output_file = self.output_manager.setup_io_files(output_dir)
+
+        _create_large_input_file(input_file, num_samples=1)
+
+        cmd = self.cmd_builder.build_generate_cmd(output_dir, input_file, retry_strategy)
+        subprocess.run(cmd, shell=True, check=True)
+
+        assert self.validator.validate_generation_output(output_file), "Output file not found"
+
+
+# Initialize test suite
+test_suite = ContextRetryTestSuite()
+
+
+@pytest.mark.gpu
```

Comment:
would be good to test sglang / trtllm / api models as well. Also would be good to test with ++use_completions_api=True in at least some cases